File: OpenCV/Binarization/naive.cpp
#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

int main(int argc, char** argv)
{

	const char* keys = {
		"{help h usage | | Help text}"
		"{@image | | Image for filtering}"
	};

	cv::CommandLineParser parser(argc, argv, keys);
	parser.about("Vivek's Binarizer");

	if (parser.has("help")) {
		parser.printMessage();
		return 0;
	}

	std::string file = parser.get<std::string>(0);

	if (!parser.check()) {
		parser.printErrors();
		return 0;
	}

	if (file == "") {
		std::cout << "Please supply a file name\n";
		return 0;
	}

	std::cout << "Loading...\n";

	cv::Mat image = cv::imread(file, 0);
	cv::imshow("Original image", image);
	std::cout << "Done!\n";

	cv::Mat hist;
	int channels[] = {0};
	int range[] = {0,256};
	int bins[] = {10};
	cv::calcHist(&image, 1, channels, cv::Mat(), hist, 1, bins, range);

	cv::waitKey(0);
	return 0;
}

File: OpenCV/Binarization/naive.py
import sys
import cv2
from matplotlib import pyplot as plt
import numpy as np

img = cv2.imread(sys.argv[1], 0)

cv2.imshow("Original Image", img)

hist = cv2.calcHist([img], [0], None, [256], [0, 256])

p1 = plt.figure(1)
plt.plot(hist)

cv2.add(img, 40, img)
np.clip(img, 0, 255, out=img)
cv2.imshow("New Image", img)

hist = cv2.calcHist([img], [0], None, [256], [0, 256])

p2 = plt.figure(2)
plt.plot(hist)

p1.show()
p2.show()

cv2.waitKey(0)
print('Done')
